### PR TITLE
 feat(edge): support edge stacks for kubernetes

### PR DESCRIPTION
**Pull request by chiptus**
needs to be merged with portainer PR (https://github.com/portainer/portainer-ee/pull/378)

blocked by #175

### CODE REVIEW

File: exec/kubernetes_deploy.go [hunk 1]
```diff
@@ -2,35 +2,90 @@ package exec
...
+		command = path.Join(binaryPath, "kubectl.exe")
+	}
+
 	return &KubernetesDeployer{
-		binaryPath: binaryPath,
+		command: command,
 	}
 }
 
 // Deploy will deploy a Kubernetes manifest inside a specific namespace
```
Comment (cmenginnz) [inline exec/kubernetes_deploy.go]: Only inside the default namespace?

File: exec/kubernetes_deploy.go [hunk 2]
```diff
@@ -2,35 +2,90 @@ package exec
...
 // it will use kubectl to deploy the manifest.
-func (deployer *KubernetesDeployer) Deploy(token, data string, namespace string) ([]byte, error) {
-	command := path.Join(deployer.binaryPath, "kubectl")
-	if runtime.GOOS == "windows" {
-		command = path.Join(deployer.binaryPath, "kubectl.exe")
+// kubectl uses in-cluster config.
+func (deployer *KubernetesDeployer) Deploy(name string, filePaths []string, prune bool) error {
+	if len(filePaths) == 0 {
```
Comment (cmenginnz) [inline exec/kubernetes_deploy.go:35]: This function always runs kubectl with in-cluster config which has admin permission. It's a security hold.
As DeployRawConfig does,  it should run kubectl with the token from the request if there is any.

I'm not sure portainer allows non-admin users to deploy stacks or not. If the answer is not, the above comment is not necessary. But it's always good to use the token from request IMO.

File: exec/kubernetes_deploy.go [hunk 3]
```diff
@@ -2,35 +2,90 @@ package exec
...
+	_, err := cmd.Output()
+	if err != nil {
+		return fmt.Errorf("failed deploying kubernetes stack %w: %s", err, stderr.String())
 	}
 
+	return nil
+}
+
+func (deployer *KubernetesDeployer) Remove(name string, filePaths []string) error {
+	if len(filePaths) == 0 {
```
Comment (cmenginnz) [inline exec/kubernetes_deploy.go:55]: Same security concern as Deploy method.

File: exec/kubernetes_deploy.go [hunk 4]
```diff
@@ -2,35 +2,90 @@ package exec
...
+	}
+
+	return nil
+}
+
+// DeployRawConfig will deploy a Kubernetes manifest inside a specific namespace
+// it will use kubectl to deploy the manifest and receives a raw config.
+// kubectl uses in-cluster config.
+func (deployer *KubernetesDeployer) DeployRawConfig(token, config string, namespace string) ([]byte, error) {
 	args := make([]string, 0)
```
Comment (cmenginnz) [inline exec/kubernetes_deploy.go]: DeployRawConfig and Deploy share many common codes. It would be nice if pull them into a sharing method.

**Review by mcpacino — CHANGES_REQUESTED** [review]

File: internal/edge/stack.go [hunk 5]
```diff
@@ -54,16 +53,25 @@ const (
...
 type StackManager struct {
-	isSwarm            *bool
-	stacks             map[edgeStackID]*edgeStack
-	stopSignal         chan struct{}
-	dockerStackService agent.DockerStackService
-	portainerURL       string
-	endpointID         string
-	isEnabled          bool
-	httpClient         *client.PortainerClient
+	engineStatus *engineStatus
```
Comment (cmenginnz) [inline internal/edge/stack.go]: It looks like that using type engineStatus but pointer type *engineStatus is enough and simpler here.
And naming engineType makes more sense to me.

**Review by mcpacino — CHANGES_REQUESTED** [review]

File: exec/kubernetes_deploy.go (hunk 1 above)
Comment (chiptus) [reply exec/kubernetes_deploy.go]: yes

File: exec/kubernetes_deploy.go (hunk 2 above)
Comment (chiptus) [reply exec/kubernetes_deploy.go:35]: edge stacks are deployable by admin only. `Deploy` isn't called by an http request so there's no token.

File: internal/edge/stack.go [hunk 6]
```diff
@@ -53,18 +53,18 @@ const (
...
+	_ engineType = iota
+	engineTypeDockerStandalone
+	engineTypeDockerSwarm
+	engineTypeKubernetes
 )
 
 // StackManager represents a service for managing Edge stacks
 type StackManager struct {
-	engineStatus *engineStatus
+	engineType   *engineType
```
Comment (cmenginnz) [inline internal/edge/stack.go]: why need pointer here?

**Review by mcpacino — CHANGES_REQUESTED** [review]

File: internal/edge/stack.go (hunk 6 above)
Comment (chiptus) [reply internal/edge/stack.go]: to be able to have a null value (engine is not set yet)

Comment (cmenginnz) [reply internal/edge/stack.go]: It's an ok option if it needs to be a null value.

But personally I would like to explicit define the null value as a enum number. For example:
const (
	engineTypeNone engineType = iota
	engineTypeDockerStandalone
	engineTypeDockerSwarm
	engineTypeKubernetes
)

**Review by mcpacino — APPROVED** [review]

File: internal/edge/stack.go (hunk 6 above)
Comment (chiptus) [reply internal/edge/stack.go]: done

**Pull request merged**